packages/babel-plugin-transform-react-jsx-self/src/index.js

```diff
@@ -16,11 +17,82 @@ import { types as t } from "@babel/core";
 
 const TRACE_ID = "__self";
 
+/**
+ * Finds the closest parent function that provides `this`. Specifically, this looks for
+ * the first parent function that isn't an arrow function.
+ *
+ * Derived from `Scope#getFunctionParent`
+ */
+function getThisFunctionParent(path) {
+  let scope = path.scope;
+  do {
+    if (
+      scope.path.isFunctionParent() &&
+      !scope.path.isArrowFunctionExpression()
+    ) {
+      return scope.path;
+    }
+  } while ((scope = scope.parent));
+  return null;
+}
+
+/**
+ * Returns whether `this` is allowed at given path.
+ */
+function containsSuperCall(sourcePath) {
+  // We don't want to go into inner classes as their super() calls aren't relevant for us.
+  if (sourcePath.isClassDeclaration() || sourcePath.isClassExpression()) {
```

Comment:
It's enough to check if the class has a superclass (`!!sourcePath.node.superClass`), without traversing it looking for `super()` calls.

Reply:
Also, we should add this test:
```js
class A extends B {
  constructor() {
    return <Foo />;
  }
}
```

It's valid JS code even without a `super()` call, but we cannot reference `this`.

Reply:
Right, that makes sense. Also, since the `sourcePath` here will be `ClassMethod`s, is it fine to check `sourcePath.parentPath.parentPath.node.superClass !== null`?

Reply:
@nicolo-ribaudo What should be the expected output for this test?

Reply:
`sourcePath.parentPath.parentPath.node.superClass` should be ok. If you want, you can also add something
```js
t.assertClass(sourcePath.parentPath.parentPath.node)
```

> What should be the expected output for this test?

It shouldn't inject `__self: this`.

Reply:
Ok, I didn't see your explanation below the code. To be honest, I have never seen this syntax before. Anyways, I have an idea for weird cases like this. We can potentially replace `this` with `(() => { try { return this; } catch(error) { return void 0; } })()`.

Reply:
Ok nevermind. I just realized your approach already covered cases like this.